Error in building ExaCA with external JSON

Dear Developers,

I encountered an error when I followed the instructions on the readme. Here is the code I used in the terminal.

### Building with external JSON

```sh
# Change this path to desired JSON installation location
export JSON_INSTALL_DIR=`pwd`/install/json

cd json
# Configure json
cmake \
  -B build \
  -D CMAKE_BUILD_TYPE="Release" \
  -D CMAKE_INSTALL_PREFIX=$JSON_INSTALL_DIR \
  -D JSON_BuildTests=OFF
# Build json
cmake --build build
# Install json
cmake --install build
cd ../
```

### Build ExaCA

```sh
# Change this path to desired ExaCA installation location
export EXACA_INSTALL_DIR=`pwd`/install/exaca
# Change this path to Kokkos installation location
export KOKKOS_INSTALL_DIR=`pwd`/install/kokkos

# Change this path to ExaCA source
# Configure ExaCA
cd ./ExaCA
cmake \
  -B build \
  -D CMAKE_BUILD_TYPE="Release" \
  -D CMAKE_PREFIX_PATH=$KOKKOS_INSTALL_DIR;$JSON_INSTALL_DIR \
  -D CMAKE_INSTALL_PREFIX=$EXACA_INSTALL_DIR
# Build ExaCA
cmake --build build
# Install ExaCA
cmake --install build
cd ../
```

### The error while executing this code
```
-- The CXX compiler identification is GNU 9.4.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Enabled Kokkos devices: CUDA;SERIAL
CMake Warning at /root/kokkos/build/install/lib/cmake/Kokkos/KokkosConfigCommon.cmake:31 (MESSAGE):
  The installed Kokkos configuration does not support CXX extensions.
  Forcing -DCMAKE_CXX_EXTENSIONS=Off
Call Stack (most recent call first):
  /root/kokkos/build/install/lib/cmake/Kokkos/KokkosConfig.cmake:56 (INCLUDE)
  CMakeLists.txt:17 (find_package)


-- kokkos_launch_compiler is enabled globally. C++ compiler commands with -DKOKKOS_DEPENDENCE will be redirected to the appropriate compiler for Kokkos
-- Found MPI_CXX: /usr/lib/x86_64-linux-gnu/libmpichcxx.so (found version "3.1") 
-- Found MPI: TRUE (found version "3.1")  
[ 11%] Creating directories for 'json-populate'
[ 22%] Performing download step (download, verify and extract) for 'json-populate'
-- Downloading...
   dst='/root/ExaCA013/build/_deps/json-subbuild/json-populate-prefix/src/json.tar.xz'
   timeout='none'
   inactivity timeout='none'
-- Using src='https://github.com/nlohmann/json/releases/download/v3.11.2/json.tar.xz'
CMake Error at json-subbuild/json-populate-prefix/src/json-populate-stamp/download-json-populate.cmake:170 (message):
  Each download failed!

    error: downloading 'https://github.com/nlohmann/json/releases/download/v3.11.2/json.tar.xz' failed
          status_code: 35
          status_string: "SSL connect error"
          log:
          --- LOG BEGIN ---


  ALPN, offering http/1.1

  [5 bytes data]

  TLSv1.3 (OUT), TLS handshake, Client hello (1):

  [512 bytes data]

  [5 bytes data]

  error:1408F10B:SSL routines:ssl3_get_record:wrong version number

  Closing connection 0

  

          --- LOG END ---
          
    


CMakeFiles/json-populate.dir/build.make:98: recipe for target 'json-populate-prefix/src/json-populate-stamp/json-populate-download' failed
make[2]: *** [json-populate-prefix/src/json-populate-stamp/json-populate-download] Error 1
CMakeFiles/Makefile2:82: recipe for target 'CMakeFiles/json-populate.dir/all' failed
make[1]: *** [CMakeFiles/json-populate.dir/all] Error 2
Makefile:90: recipe for target 'all' failed
make: *** [all] Error 2

CMake Error at /usr/local/share/cmake-3.22/Modules/FetchContent.cmake:1087 (message):
  Build step for json failed: 2
Call Stack (most recent call first):
  /usr/local/share/cmake-3.22/Modules/FetchContent.cmake:1216:EVAL:2 (__FetchContent_directPopulate)
  /usr/local/share/cmake-3.22/Modules/FetchContent.cmake:1216 (cmake_language)
  /usr/local/share/cmake-3.22/Modules/FetchContent.cmake:1259 (FetchContent_Populate)
  CMakeLists.txt:28 (FetchContent_MakeAvailable)


-- Configuring incomplete, errors occurred!

```


This seems to be due to ```find_package(nlohmann_json 3.10.0 QUIET)``` in ```CMakeLists.txt```. When json is not detected, it will directly enter the subsequent download process without reporting any error message. Because there is a problem with the network in my environment, an error occurs and the installation cannot be performed. What's more, ```CMAKE_PREFIX_PATH=$KOKKOS_INSTALL_DIR;$JSON_INSTALL_DIR``` in ```Readme.md``` could mistakenly use ```;```, which might be replaced by ```CMAKE_PREFIX_PATH=$KOKKOS_INSTALL_DIR``` and ```nlohmann_json_DIR=$JSON_INSTALL_DIR```

I changed ```CMakeLists.txt``` by using a flag ```ExaCA_REQUIRE_External_JSON``` and ```find_package(nlohmann_json 3.10.0 REQUIRED)``` to get the error message. I added ```-D ExaCA_REQUIRE_External_JSON=ON``` in cmake. The cmake instruction in the terminal is as follows: 
```sh
cmake \
  -B build \
  -D CMAKE_BUILD_TYPE="Release" \
  -D CMAKE_PREFIX_PATH=$KOKKOS_INSTALL_DIR \
  -D CMAKE_INSTALL_PREFIX=$EXACA_INSTALL_DIR \
  -D ExaCA_REQUIRE_External_JSON=ON \
  -D nlohmann_json_DIR=$JSON_INSTALL_DIR
```

### My modified ```CMakeLists.txt```

```
cmake_minimum_required(VERSION 3.12)

project(ExaCA LANGUAGES CXX)
set(PROJECT_VERSION "1.4.0-dev")

include(GNUInstallDirs)

option(ExaCA_REQUIRE_FINCH "Build with Finch heat transport support" OFF)
# Build with Finch if available or if requested.
find_package(Finch QUIET)
if(ExaCA_REQUIRE_FINCH OR Finch_FOUND)
  set(ExaCA_ENABLE_FINCH ON)
  find_package(Finch REQUIRED COMPONENTS Finch::Core Finch::ScanPaths)
  message( STATUS "Finch heat transfer enabled." )
endif()

find_package(Kokkos 4.0 REQUIRED)
find_package(MPI REQUIRED)

if (${CMAKE_VERSION} VERSION_GREATER_EQUAL "3.24")
  cmake_policy(SET CMP0135 NEW)
endif()

# ---------------Where I changed----------------------------------------------------------------------

option(ExaCA_REQUIRE_External_JSON "Build with external JSON" OFF)

if(ExaCA_REQUIRE_External_JSON)
  find_package(nlohmann_json 3.10.0 REQUIRED)
endif()

if(NOT ExaCA_REQUIRE_External_JSON)
# ---------------Where I changed----------------------------------------------------------------------
  include(FetchContent)
  # Using most recent release here
  FetchContent_Declare(json URL https://github.com/nlohmann/json/releases/download/v3.11.2/json.tar.xz)
  FetchContent_MakeAvailable(json)
  # Needs to be installed only in this case.
  install(TARGETS nlohmann_json
    EXPORT ExaCA_Targets
    LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
    ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
endif()

find_package(Git)
if(GIT_FOUND AND IS_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/.git)
  EXECUTE_PROCESS(
    COMMAND           ${GIT_EXECUTABLE} log --pretty=format:%H -n 1
    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
    OUTPUT_VARIABLE   ExaCA_GIT_COMMIT_HASH
    )
else()
  set(ExaCA_GIT_COMMIT_HASH "Not a git repository")
endif()
message(STATUS "ExaCA commit: ${ExaCA_GIT_COMMIT_HASH}")

# NOTE: this order is specifically used in the unit tests.
# FIXME: remove PTHREAD
set(EXACA_SUPPORTED_DEVICES SERIAL PTHREAD THREADS OPENMP CUDA HIP)

add_subdirectory(src)

configure_file(${CMAKE_CURRENT_SOURCE_DIR}/cmake/ExaCAconfig.cmakein
  ${CMAKE_CURRENT_BINARY_DIR}/ExaCAconfig.cmake @ONLY)
install(FILES "${CMAKE_CURRENT_BINARY_DIR}/ExaCAconfig.cmake"
  DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/ExaCA)

# Install data files
file(GLOB MATERIALS examples/Materials/*)
file(GLOB SUBSTRATE examples/Substrate/GrainOrientation*.csv)
install(FILES ${MATERIALS}
  DESTINATION ${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_DATADIR}/ExaCA)
install(FILES ${SUBSTRATE}
  DESTINATION ${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_DATADIR}/ExaCA)

option(ExaCA_ENABLE_TESTING "Build unit tests" OFF)

if(ExaCA_ENABLE_TESTING)
  find_package(GTest 1.10 REQUIRED)
  # Workaround for FindGTest module in CMake older than 3.20
  if(TARGET GTest::gtest)
    set(gtest_target GTest::gtest)
  elseif(TARGET GTest::GTest)
    set(gtest_target GTest::GTest)
  else()
    message(FATAL_ERROR "bug in GTest find module workaround")
  endif()
  enable_testing()

  set(TEST_HARNESS_DIR ${CMAKE_SOURCE_DIR}/unit_test)
  set(EXACA_TEST_DEVICES)
  foreach(_device ${EXACA_SUPPORTED_DEVICES})
    if(Kokkos_ENABLE_${_device})
      list(APPEND EXACA_TEST_DEVICES ${_device})
    endif()
  endforeach()

  add_subdirectory(unit_test)
endif()

add_subdirectory(bin)

add_subdirectory(analysis)

# clang-format
set(CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/cmake)
find_package( CLANG_FORMAT 10)
if(CLANG_FORMAT_FOUND)
  file(GLOB_RECURSE FORMAT_SOURCES *.[c,h]pp)
  add_custom_target(format
    COMMAND ${CLANG_FORMAT_EXECUTABLE} -i -style=file ${FORMAT_SOURCES}
    DEPENDS ${FORMAT_SOURCES})
endif()

```

### Successfully obtained the correct error message

```
-- The CXX compiler identification is GNU 9.4.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Enabled Kokkos devices: CUDA;SERIAL
CMake Warning at /root/kokkos/build/install/lib/cmake/Kokkos/KokkosConfigCommon.cmake:31 (MESSAGE):
  The installed Kokkos configuration does not support CXX extensions.
  Forcing -DCMAKE_CXX_EXTENSIONS=Off
Call Stack (most recent call first):
  /root/kokkos/build/install/lib/cmake/Kokkos/KokkosConfig.cmake:56 (INCLUDE)
  CMakeLists.txt:17 (find_package)


-- kokkos_launch_compiler is enabled globally. C++ compiler commands with -DKOKKOS_DEPENDENCE will be redirected to the appropriate compiler for Kokkos
-- Found MPI_CXX: /usr/lib/x86_64-linux-gnu/libmpichcxx.so (found version "3.1") 
-- Found MPI: TRUE (found version "3.1")  
CMake Error at CMakeLists.txt:29 (find_package):
  By not providing "Findnlohmann_json.cmake" in CMAKE_MODULE_PATH this
  project has asked CMake to find a package configuration file provided by
  "nlohmann_json", but CMake did not find one.

  Could not find a package configuration file provided by "nlohmann_json"
  (requested version 3.10.0) with any of the following names:

    nlohmann_jsonConfig.cmake
    nlohmann_json-config.cmake

  Add the installation prefix of "nlohmann_json" to CMAKE_PREFIX_PATH or set
  "nlohmann_json_DIR" to a directory containing one of the above files.  If
  "nlohmann_json" provides a separate development package or SDK, be sure it
  has been installed.


-- Configuring incomplete, errors occurred!
See also "/root/ExaCA013/build/CMakeFiles/CMakeOutput.log".
```

I found that the problem seems to be that the file ```nlohmann_jsonConfig.cmake``` was not found, so I modified the corresponding path ```nlohmann_json_DIR="$JSON_INSTALL_DIR/lib/cmake/nlohmann_json"``` to ensure that it can be found. 
```sh
cmake \
  -B build \
  -D CMAKE_BUILD_TYPE="Release" \
  -D CMAKE_PREFIX_PATH=$KOKKOS_INSTALL_DIR \
  -D CMAKE_INSTALL_PREFIX=$EXACA_INSTALL_DIR \
  -D ExaCA_REQUIRE_External_JSON=ON \
  -D nlohmann_json_DIR="$JSON_INSTALL_DIR/lib/cmake/nlohmann_json"
```

Then I run ```cmake --build build```, there were some errors:
```
[ 11%] Building CXX object src/CMakeFiles/ExaCA-Core.dir/CAinfo.cpp.o
[ 22%] Building CXX object src/CMakeFiles/ExaCA-Core.dir/CAparsefiles.cpp.o
[ 33%] Linking CXX static library libExaCA-Core.a
[ 33%] Built target ExaCA-Core
[ 44%] Building CXX object bin/CMakeFiles/ExaCA.dir/run.cpp.o
[ 55%] Linking CXX executable ExaCA
[ 55%] Built target ExaCA
[ 55%] Built target ExaCA-Kokkos
[ 66%] Building CXX object analysis/src/CMakeFiles/ExaCA-Analysis.dir/GAutils.cpp.o
[ 77%] Linking CXX static library libExaCA-Analysis.a
[ 77%] Built target ExaCA-Analysis
[ 88%] Building CXX object analysis/bin/CMakeFiles/ExaCA-GrainAnalysis.dir/runGA.cpp.o
/root/ExaCA/analysis/src/GArepresentativeregion.hpp(747): error: namespace "std" has no member "setprecision"

/root/ExaCA/analysis/src/GArepresentativeregion.hpp(815): error: namespace "std" has no member "setprecision"

2 errors detected in the compilation of "/root/ExaCA/analysis/bin/runGA.cpp".
analysis/bin/CMakeFiles/ExaCA-GrainAnalysis.dir/build.make:75: recipe for target 'analysis/bin/CMakeFiles/ExaCA-GrainAnalysis.dir/runGA.cpp.o' failed
make[2]: *** [analysis/bin/CMakeFiles/ExaCA-GrainAnalysis.dir/runGA.cpp.o] Error 1
CMakeFiles/Makefile2:275: recipe for target 'analysis/bin/CMakeFiles/ExaCA-GrainAnalysis.dir/all' failed
make[1]: *** [analysis/bin/CMakeFiles/ExaCA-GrainAnalysis.dir/all] Error 2
Makefile:135: recipe for target 'all' failed
make: *** [all] Error 2
```
I added ```#include<iomanip>``` to file ```/ExaCA/analysis/src/GArepresentativeregion.hpp``` to solve this problem. I want to know if this will cause any problems.

Therefore, my issues can be summarized as follows:

1. ```find_package(nlohmann_json 3.10.0 QUIET)``` in ```CMakeLists.txt``` may cause the error message about "the local JSON library cannot be found" ignored. Could we consider introducing a new flag ```ExaCA_REQUIRE_External_JSON```, as shown in My modified ```CMakeLists.txt```, to ensure that local JSON is not downloaded from the network during installation?
2. There seems to be an issue with the cmake command ```-D CMAKE_PREFIX_PATH=$KOKKOS_INSTALL_DIR;$JSON_INSTALL_DIR``` for external JSON installation in readme, and ```$JSON_INSTALL_DIR``` cannot get ```nlohmann_jsonConfig.cmake```. Can we consider changing it to the following?
```sh
cmake \
  -B build \
  -D CMAKE_BUILD_TYPE="Release" \
  -D CMAKE_PREFIX_PATH=$KOKKOS_INSTALL_DIR \
  -D CMAKE_INSTALL_PREFIX=$EXACA_INSTALL_DIR \
  -D ExaCA_REQUIRE_External_JSON=ON \
  -D nlohmann_json_DIR="$JSON_INSTALL_DIR/lib/cmake/nlohmann_json"
```
3. Is it appropriate to add ```#include<iomanip>``` to file ```/ExaCA/analysis/src/GArepresentativeregion.hpp```?

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -20,17 +20,22 @@ find_package(MPI REQUIRED)
 if (${CMAKE_VERSION} VERSION_GREATER_EQUAL "3.24")
   cmake_policy(SET CMP0135 NEW)
 endif()
-find_package(nlohmann_json 3.10.0 QUIET)
-if(NOT NLOHMANN_JSON_FOUND)
-  include(FetchContent)
-  # Using most recent release here
-  FetchContent_Declare(json URL https://github.com/nlohmann/json/releases/download/v3.11.2/json.tar.xz)
-  FetchContent_MakeAvailable(json)
-  # Needs to be installed only in this case.
-  install(TARGETS nlohmann_json
-    EXPORT ExaCA_Targets
-    LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
-    ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
+option(ExaCA_REQUIRE_EXTERNAL_JSON "Build without automatic JSON dependency download" OFF)
+if(ExaCA_REQUIRE_EXTERNAL_JSON)
+  find_package(nlohmann_json 3.10.0 REQUIRED)
+else()
+  find_package(nlohmann_json 3.10.0 QUIET)
+  if(NOT NLOHMANN_JSON_FOUND)
+    include(FetchContent)
+    # Using most recent release here
+    FetchContent_Declare(json URL https://github.com/nlohmann/json/releases/download/v3.11.2/json.tar.xz)
+    FetchContent_MakeAvailable(json)
+    # Needs to be installed only in this case.
+    install(TARGETS nlohmann_json
+      EXPORT ExaCA_Targets
+      LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+      ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
+  endif()
 endif()
 
 find_package(Git)

